---
title: "Differential Expression by Polarization Method"
author: "Timothy Smyth"
date: "2024-03-27"
output: html_document
---

# Differential Expression by Polarization State

### This file performs differential expression analysis of monocyte-derived macrophage count data by polarization method using limma-voom. Raw count data of gene and sample filtered MDM data from 'Gene and Sample Filtering' is the input data for this analysis.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Environment setup

```{r message = FALSE}
rm(list = ls(all.names = TRUE)) # clears global environ.

options(java.parameters = "-Xmx8g" )

# Load packages
library(tidyverse) # for data cleaning
library(dplyr) 
library(broom)
library(tibble)
library(edgeR)
library(limma)
library(EnhancedVolcano)
library(AnnotationHub)
library(GO.db)
library(biomaRt)
library(topGO)
```

### Load and format count data

```{r}
# This is gene filtered and sample selected but not limma-voom corrected data
# The model written below takes batch (series) into account
load("Selected_Count_Data.RData")

# Isolate sample info and set to factor
info$polarization <- factor(info$polarization, 
                            levels = c("M0", "M1", "M2"))

series_id <- factor(info$series_id)

polarization_method <- factor(info$ID,
                              level = c ("M0", 
                                         "LPS_IFN", "LPS", "IFN", 
                                         "IL4", "IL10", "IL13", "IL4_IL13"))

# Create DGEList object
dataset <- DGEList(t(as.matrix(MDM_df)), 
                   group = polarization_method)

dataset <- calcNormFactors(dataset)

# Print the dimensions of the new object (Genes x Samples)
dim(dataset)
```

### Create model matrix, run voom, and fit linear models

```{r}
# Create a model with polarization and series as factors
model <- model.matrix(~0 + polarization_method + series_id)

colnames(model)[1:8] <- c("M0", 
                          "LPS_IFN", "LPS", "IFN", 
                          "IL4", "IL10", "IL13", "IL4_IL13")

# Check if any coefficients in model cannot be estimated
# Return value of "NULL" means model has full column rank
nonEstimable(model.matrix(~0 + polarization_method + series_id))

# Run voom and plot mean-variance trend
# voom function automatically converts counts to logCPM
# by extracting library size and calculating normalization factors
y <- voom(dataset, 
          design = model, 
          plot = T)

# Fit linear model using weighted least squares for genes
fit <- lmFit(y, 
             model)
```

### Calculate differentially expressed genes between groups as defined by makeContrasts using contrast.fit, perform empirical Bayes smoothing to gene-wise standard errors with eBayes, and present statistics using topTable

```{r}
# Make a list of contrasts
list <- list(makeContrasts(LPS_IFN-LPS, 
                           levels = model),
             
             makeContrasts(LPS_IFN-IFN, 
                           levels = model),
             
             makeContrasts(LPS_IFN-M0, 
                           levels = model),
             
             #############
             
             makeContrasts(IFN-LPS, 
                           levels = model), 
             
             
             makeContrasts(IFN-M0, 
                           levels = model),
             
             #############
             
             makeContrasts(LPS-IFN, 
                           levels = model), 
             
             makeContrasts(LPS-M0, 
                           levels = model),
             
             #############
             
             makeContrasts(IL4_IL13-IL4, 
                           levels = model), 
             
             makeContrasts(IL4_IL13-IL10, 
                           levels = model), 
             
             makeContrasts(IL4_IL13-IL13, 
                           levels = model),
             
             makeContrasts(IL4_IL13-M0, 
                           levels = model),
             
             #############
             
             makeContrasts(IL4-IL10, 
                           levels = model), 
             
             makeContrasts(IL4-M0, 
                           levels = model), 
             
             #############
             
             makeContrasts(IL4-IL13, 
                           levels = model), 
             
             makeContrasts(IL13-M0, 
                           levels = model), 
             
             #############
             
             makeContrasts(IL10-IL13, 
                           levels = model),
             
             makeContrasts(IL10-M0, 
                           levels = model))

names <- c('M(LPS_IFN) vs M(LPS)', 'M(LPS_IFN) vs M(IFN)', 'M(LPS_IFN)',
           'M(IFN) vs M(LPS)', 'M(IFN)', 
           'M(LPS) vs M(IFN)', 'M(LPS)',
           'M(IL4_IL13) vs M(IL4)', 'M(IL4_IL13) vs M(IL10)', 'M(IL4_IL13) vs M(IL13)', 'M(IL4_IL13)',
           'M(IL4) vs M(IL10)', 'M(IL4)',
           'M(IL4) vs M(IL13)', 'M(IL13)', 
           'M(IL10) vs M(IL13)', 'M(IL10)')

# Calculate DEGs
DEGs <- lapply(list, function(x){
  
  tmp <- contrasts.fit(fit, x)
  tmp <- eBayes(tmp)
  
  topTable(tmp, sort.by = "P", n = Inf)
  
})

# Results are the following information:

# logFC: log2 fold change of group1/group2
# AveExpr: Average expression across all samples, in log2 CPM
# t: logFC divided by its standard error
# P.Value: Raw p-value (based on t) from test that logFC differs from 0
# adj.P.Val: Benjamini-Hochberg false discovery rate adjusted p-value
# B: log-odds that gene is DE

# Isolate DEGs
DEG_Results <- as.list(1:length(DEGs))

DEG_Results <- lapply(DEG_Results, function(x){
  
  tmp <- data.frame(cbind(DEGs[[x]][["logFC"]],
                          DEGs[[x]][["P.Value"]],
                          DEGs[[x]][["adj.P.Val"]],
                          DEGs[[x]]["AveExpr"]),
                    row.names = rownames(DEGs[[x]])) 
  
  colnames(tmp) <- c('log2FC', 'P.Value', 'adj.P.Val', 'Expression')
  
  tmp
  
})

names(DEG_Results) <- names
```

### Align HUGO gene IDs with Ensembl gene IDs and generate volcano plots for each comparison

```{r echo = T, results = 'hide'}
ah <- AnnotationHub()
query(ah, "EnsDb.Hsapiens.v107")
edb <- ah[["AH104864"]]
txs <- transcripts(edb, 
                   columns = c("tx_id", 
                               "tx_biotype", 
                               "tx_id_version", 
                               "gc_content", 
                               "gene_name", 
                               "gene_id"))

Genes <- data.frame(txs@elementMetadata@listData[["gene_name"]], 
                    txs@elementMetadata@listData[["gene_id"]])

names(Genes) <- c('gene_name', 'gene_id')

Genes <- Genes[!duplicated(Genes$gene_id), ] 

named_genes <- Genes %>% subset(gene_name != '')
unnamed_genes <- Genes %>% subset(gene_name == '')
unnamed_genes$gene_name <- unnamed_genes$gene_id

Genes <- rbind(named_genes, unnamed_genes)

Symbol <- data.frame(Genes$gene_name, row.names = Genes$gene_id)

volcano <- list()

# Make these in a for loop for easy graph naming
for(i in 1:length(DEG_Results)){
  
  # Determine the number of significantly upregulated genes
  up <- DEG_Results[[i]] %>% count(log2FC > 2 &
                                     adj.P.Val < 0.05)
  
  up <- as.numeric(up[2, 2])
  
  # Determine the number of significantly downregulated genes
  down <- DEG_Results[[i]] %>% count(log2FC < -2 &
                                       adj.P.Val < 0.05)
  
  down <- as.numeric(down[2, 2])
  
  DEG_Results[[i]] <- DEG_Results[[i]] %>% arrange(desc(log2FC))
  DEG_Results[[i]]$HUGO_IDs <- Symbol[rownames(DEG_Results[[i]]), ]
  DEG_Results[[i]]$Top_Genes <- Symbol[rownames(DEG_Results[[i]]), ]
  
  selectLab_up <- DEG_Results[[i]] %>% 
    subset(log2FC >= 2 & adj.P.Val < 0.05) %>% 
    arrange(desc(log2FC)) %>% 
    rownames()
  
  selectLab_up <- selectLab_up[1:10] %>% na.omit()

  selectLab_down <- DEG_Results[[i]] %>% 
    subset(log2FC <= -2 & adj.P.Val < 0.05) %>% 
    arrange(log2FC) %>% 
    rownames()
  
  selectLab_down <- selectLab_down[1:10] %>% na.omit()
  
  selectLabs <- c(selectLab_up, selectLab_down)
 
  DEG_Results[[i]][!rownames(DEG_Results[[i]]) %in% selectLabs, 6] <- ''

  volcano[[i]] <- EnhancedVolcano(DEG_Results[[i]],
                                  lab = DEG_Results[[i]]$Top_Genes,
                                  labSize = 3.5,
                                  drawConnectors = TRUE,
                                  widthConnectors = 0.5,
                                  max.overlaps = Inf,
                                  maxoverlapsConnectors = Inf,
                                  boxedLabels = TRUE,
                                  x = 'log2FC',
                                  y = 'adj.P.Val',
                                  FCcutoff = 2,
                                  pCutoff = 0.05,
                                  title = names(DEG_Results)[[i]],
                                  pointSize = 1) 
  
  # Determine the y axis limit
  max <- max(volcano[[i]][["plot_env"]][["ylim"]]) * 1.1
  
  # Add arrows and numbers describing the number of up or downregulated genes
  volcano[[i]] <- volcano[[i]] + 
    
    # Set x axis from -15 to 15 and extend y axis
    ggplot2::coord_cartesian(xlim = c(-15, 15), 
                             ylim = c(0, max)) +
    
    # Set tick marks for x axis
    ggplot2::scale_x_continuous(breaks = seq(-15, 15, by = 5)) +
    
    # Place downregulated numbers on plot
    ggplot2::geom_text(x = -12.5,
                       y = max* 0.95,
                       label = paste0(down),
                       size = 8.5) +
    
    # Place downregulation arrow on plot
    ggplot2::geom_segment(x = -15,
                          y = max,
                          xend = -15,
                          yend = max * 0.9,
                          arrow = arrow(length = unit(2, "mm")), 
                          linewidth = 1) +
    
    # Place upregulated numbers on plot
    ggplot2::geom_text(x = 13.5,
                       y = max * 0.95,
                       label = paste0(up),
                       size = 8.5) + 
    
    # Place upregulation arrow on plot
    ggplot2::geom_segment(x = 11,
                          y = max * 0.9,
                          xend = 11,
                          yend = max,
                          arrow = arrow(length = unit(2, "mm")), 
                          linewidth = 1)
  
  # Save the resulting plot
  tiff(paste0(names(DEG_Results)[[i]], '.tiff'),
       units = "cm",
       width = 18,
       height = 16,
       res = 300)

  print(volcano[[i]])

  dev.off()
  
}

Method_DEG_Results <- DEG_Results

save(Method_DEG_Results,
     Symbol,
     file = 'Method_DEG_Results.RData')
```

### Save the DEG results

```{r}
# Isolate genes with adj.P.Val < 0.05 for IPA
DEGs <- lapply(DEG_Results, function(x){
  
  tmp <- x %>% subset(adj.P.Val < 0.05 & (log2FC >= 2 | log2FC <= -2))
  
  # Change . back to - for gene names as IPA expects -
  row.names(tmp) <- gsub('\\.', "\\-", row.names(tmp))
  
  tmp
  
})

names(DEGs) <- names

# Export DEGs [Log2FC and BH adjusted p values]
lapply(names(DEGs), function(x){
  
  write.csv(DEGs[[x]], file = paste0(x, '.csv'))
  
})
```